content/courses/Ada_For_The_C_Embedded_Developer/chapters/05_SPARK.rst

```diff
@@ -7,36 +7,105 @@ Enhancing Verification with SPARK and Ada
-languages, where the exception is being propagated from call frame to call
-frame up until it either reaches the main subprogram (terminating the
-application) or is caught by an so called exception handler in one of the call
-frames. For example:
+Understanding Exceptions and Dynamic Checks
+-------------------------------------------
+
+In Ada, several common programming errors that are not already detected 
+at compile-time are detected instead at run-time, triggering 
+"exceptions" that interrupt the normal flow of execution. For example, 
+an exception is raised by an attempt to access an array component via an 
+index that is out of bounds, thereby precluding exploits based on buffer 
+overflow. Several other cases also raise language-defined exceptions, 
+such as scalar range constraint violations, and users may declare and 
+raise their own application-specific exceptions. 
+
+We will loosely define a "frame" as a language construct that has a call
+stack entry when called, for example a procedure or function body. There
+are a few other constructs that are also pertinent but this definition
+will suffice for now.
+
+Frames can have "exception handlers" that specify the response to be
+executed when an exception is "raised." These handlers can be specified to
+"handle" one or more specific exceptions, thus allowing the response to
+be tailored to specific exceptions or generalized to more than one.
+
+Structurally, the exception handlers are grouped together and textually
+separated from the rest of the code in a frame. As a result, there is a
+sequence of statements representing the normal flow of execution,
+distinct from another section representing the error handling. The
+reserved word "exception" separates these two sections. This separation
+helps simplify the overall flow, increasing understandability. In
+particular, status result codes are not required so there is no mixing
+of error checking and normal progress. If no exception is raised the
+exception handler section is automatically skipped during execution.
+
+For example, in the following procedure line 4 separates the two
+sections. Lines 5 through 10 contain the error handling part,
+consisting in this case of three exception handlers:
+
+:code-config:`accumulate_code=True`
+
+.. code:: ada project=Courses.Ada_For_C_Embedded_Dev.SPARK.Exceptions
+
+    procedure P is
+    begin
+       Statements_That_Might_Raise_Exceptions;
+    exception
+       when Exception_A =>
+          Handle_Exception_A;
+       when Exception_B =>
+          Handle_Exception_B;
+       when Exception_C =>
+          Handle_Exception_C;
+    end P;
+
+The name :ada:`Exception_A` refers to a specific visible exception
+declared somewhere with that name, as do the other two exceptions
+referenced. Each handler consists of an arbitrary sequence of
+statements, in this case specific procedures called in response to those
+specific exceptions.
```

Comment:
Another suggestion:
```suggestion
specific exceptions. For example, when handling :ada:`Exception_A`, we call :ada:`Handle_Exception_A`.
```
I know, this is pretty obvious when reading the code. Sometimes, however, it's better to add those details just to ensure all readers are understanding the context.